src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateGroup.cs

```diff
@@ -0,0 +1,283 @@
+                                                               && x.InputParameterName.Equals(parameter.InputParameterName, StringComparison.OrdinalIgnoreCase))))
+                {
+                    invalidParameterList.Add(new InvalidParameterInfo(
+                                                InvalidParameterInfo.Kind.InvalidParameterValue,
+                                                parameter.InputParameterFormat,
+                                                parameter.ParameterValue,
+                                                parameter.InputParameterName));
+                }
+            }
+
+            return invalidParameterList;
+        }
+
+        /// <summary>
+        /// The method returns the single invokable template with highest precedence
+        /// </summary>
+        /// <param name="highestPrecedenceTemplate">Contains the invokable template with highest precedence</param>
+        /// <param name="useDefaultLanguage">Defines if default language template should be preferred in case of ambiguity</param>
+        /// <returns>
+        /// <see cref="true"/> when single invokable template with highest precedence can be defined, 
+        /// <see cref="false"/> otherwise
+        /// </returns>
+        internal bool TryGetHighestPrecedenceInvokableTemplate(out ITemplateMatchInfo highestPrecedenceTemplate, bool useDefaultLanguage = false)
+        {
+            highestPrecedenceTemplate = null;
+            if (!InvokableTemplates.Any())
+            {
+                return false;
+            }
+            IEnumerable<ITemplateMatchInfo> highestPrecendenceTemplates = GetHighestPrecedenceInvokableTemplates(useDefaultLanguage);
+            if (highestPrecendenceTemplates.Count() == 1)
+            {
+                highestPrecedenceTemplate = highestPrecendenceTemplates.First();
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// The method returns the invokable templates with highest precedence
+        /// </summary>
+        /// <param name="useDefaultLanguage">Defines if default language template should be preferred in case of ambiguity</param>
+        /// <returns>
+        /// the enumerator of invokable templates with highest precedence
+        /// </returns>
+        internal IEnumerable<ITemplateMatchInfo> GetHighestPrecedenceInvokableTemplates(bool useDefaultLanguage = false)
+        {
+            IEnumerable<ITemplateMatchInfo> highestPrecedenceTemplates;
+            if (!InvokableTemplates.Any())
+            {
+                return new List<ITemplateMatchInfo>();
+            }
+
+            int highestPrecedence = InvokableTemplates.Max(t => t.Info.Precedence);
+            highestPrecedenceTemplates = InvokableTemplates.Where(t => t.Info.Precedence == highestPrecedence);
+
+            if (useDefaultLanguage && highestPrecedenceTemplates.Count() > 1)
```

Comment:
minor: consider Any instead of Count for IEnumerable

https://stackoverflow.com/questions/305092/which-method-performs-better-any-vs-count-0

Reply:
Since criteria here is >1 Any cannot be used.